java/hive/src/main/java/io/v6d/hive/ql/io/VineyardInputFormat.java

```diff
@@ -156,63 +178,50 @@ class VineyardRecordReader implements RecordReader<NullWritable, RecordWrapperWr
-                byte[] buffer = new byte[(int) fileStatus.getLen()];
-                int len = in.read(buffer, 0, (int) fileStatus.getLen());
-                if (len <= 0) {
-                    continue;
-                }
-                String[] objectIDs = new String(buffer, StandardCharsets.US_ASCII).split("\n");
-                for (val objectID : objectIDs) {
-                    try {
-                        ObjectID tableID = ObjectID.fromString(objectID);
-                        Table table =
-                                (Table)
-                                        ObjectFactory.getFactory()
-                                                .resolve(client.getMetaData(tableID));
-                        for (val batch : table.getBatches()) {
-                            recordTotal += batch.getRowCount();
-                            batch.setResolver(resolver);
-                            this.batches[this.recordBatchIndex++] = batch;
-                        }
-                        schema = table.getSchema().getSchema();
-                    } catch (ObjectNotExists | NumberFormatException e) {
-                        // Skip some invalid file.
-                        Context.println(
-                                "Skipping invalid file: "
-                                        + tableFilePath
-                                        + ", content: "
-                                        + new String(buffer, StandardCharsets.US_ASCII));
-                        break;
-                    }
-                }
-            }
+        if (tableStatus.isDirectory()) {
+            throw new IOException("Path of table is a dir!");
+        }
+
+        FSDataInputStream in = fs.open(path);
+        byte[] buffer = new byte[(int) tableStatus.getLen()];
+        int len = in.read(buffer, 0, (int) tableStatus.getLen());
+        if (len <= 0) {
+            in.close();
+            return;
+        }
+
+        // getSplits will ensure that the file referenced by path is a valid table.
+        String[] objectIDs = new String(buffer, StandardCharsets.US_ASCII).split("\n");
+        ObjectID tableID = ObjectID.fromString(objectIDs[0]);
+        Context.println("try to build table in reader");
+        ObjectMeta tableMeta = client.getMetaData(tableID, false);
+        Table table = (Table) ObjectFactory.getFactory().resolve(tableMeta);
+
+        recordTotal = split.getLength();
+        Context.println("Start:" + split.getStart() + " length:" + split.getLength());
+        for (int i = (int)split.getStart(); i < split.getLength() + split.getStart(); i++) {
+            val batch = table.getBatch(i);
+            batch.setResolver(resolver);
+            this.batches[this.recordBatchIndex++] = batch;
         }
+        schema = table.getSchema().getSchema();
         // reset to the beginning
         this.recordBatchIndex = -1;
         this.recordBatchInnerIndex = 0;
+        in.close();
```

Comment:
Use try-with-resource style for file I/O to simplify the logic for open/close.

Reply:
Done